### PR TITLE
feat(mac): install script rather than drag+drop

### PR DESCRIPTION
**Pull request by mcdurdin**
This PR updates the Mac install experience to use a simple install script. It is intended as an interim until we have the time to write a full install experience that smooths out all the nasty bumps.

The driver for this was discovering during pre-beta testing in #2526 that the drag+drop metaphor for installing into Input Methods folder aliases no longer works in Catalina.

The replacement is an app (AppleScript) that:

1. Kills any instances of Keyman already running
2. Copies the Keyman app folder over to ~/Library/Input Methods/Keyman (removing and then replacing existing Keyman app files)
3. Adds Keyman as an input source (so the user no longer needs to take this step)
4. Gives instructions on how to jump through the various Apple security hoops to complete installation.

The security hoops are presented after Keyman starts for the first time, which is significantly less than ideal. However, the aim of this PR is to get the install experience *working* again. A future feature update will be a "proper" installer (not AppleScript), which does the above steps, but also triggers the security request from Apple, hopefully opening the System Preferences dialog in the process (for a good example of what can be achieved, see the Parallels installer).

One of the niggling issues is that after enabling Keyman in the "Input Monitoring" section of the Privacy preference panel, you have to restart Keyman (which is done for you) -- but the input source menu does not switch Keyman off, so the user has to do that and then switch Keyman back on again before being able to open the menu *yet again* to change preferences.

I would much prefer to see the preferences app being a standalone app rather than embedded in the Apple Input Source preferences model, which throws a bunch of silly constraints at it. It would allow us to run it directly from the installer, for example, and make it accessible outside the input menu. But that's all grist for a future update.

### CODE REVIEW

**Comment by mcdurdin** [comment]
# User Testing:

**Requires a VM**

The point of this user testing exercise is to see how well a user new to this process can navigate the installer without instructions. So:

1. Create a new macOS VM (e.g. 10.15.2 or 10.15.3)
2. Download the dmg, and try and install -- follow the prompts.
3. Document your experiences :) 

Rinse and repeat with an older version of macOS, for verification purposes.

@MakaraSok

File: mac/setup/textinputsource/main.c [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+                    else
+                        err = TISEnableInputSource(tis);
+                    CFRelease(tis);
+                }
+                ret = tis ? (int) err : 1;
+                break;
+            case 'h':
+            case '?':
+            default:
+                fprintf(stderr, "Usage: %s %s\n\t\%s\n\t%s\n\t%s\n\t%s\n\t%s\n",
```
Comment (darcywong00) [inline mac/setup/textinputsource/main.c]:
```suggestion
                fprintf(stderr, "Usage: %s %s\n\t%s\n\t%s\n\t%s\n\t%s\n\t%s\n",
```

File: mac/Keyman4Mac/Keyman4Mac/Info.plist [hunk 2]
```diff
@@ -17,11 +17,11 @@
 	<key>CFBundlePackageType</key>
 	<string>APPL</string>
 	<key>CFBundleShortVersionString</key>
-	<string>Keyman $(PRODUCT_VERSION) for macOS</string>
+	<string>13.0.99</string>
```
Comment (darcywong00) [inline mac/Keyman4Mac/Keyman4Mac/Info.plist]: Does this string title get updated somewhere else? (with the new $(PRODUCT_VERSION))

File: mac/setup/build.sh [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+TARGET_APP_PATH="$TARGETAPP"
+ALTOOL_LOG_PATH="$TARGETPATH/altool.log"
+
+echo_heading "Zipping Install Keyman.app for notarization to $TARGET_ZIP_PATH"
+
+/usr/bin/ditto -c -k --keepParent "$TARGET_APP_PATH" "$TARGET_ZIP_PATH"
+
+echo_heading "Uploading Install Keyman.zip to Apple for notarization"
+
+xcrun altool --notarize-app --primary-bundle-id "com.Keyman.install.zip" --asc-provider "$APPSTORECONNECT_PROVIDER" --username "$APPSTORECONNECT_USERNAME" --password @env:APPSTORECONNECT_PASSWORD --file "$TARGET_ZIP_PATH" --output-format xml > $ALTOOL_LOG_PATH || fail "altool failed"
```
Comment (darcywong00) [inline mac/setup/build.sh]: Do you want to fail the build if `APPSTORECONNECT_*` credentials aren't provided?

File: mac/setup/textinputsource/main.c [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+int
+main (int argc, char * argv[])
+{
+    int ret = -1;
+    int c;
+    TISInputSourceRef tis;
+    CFStringRef name;
+    OSStatus err = 0;
+
+    while ((c = getopt(argc, argv, "s:e:d:t:h")) != -1)
```
Comment (darcywong00) [inline mac/setup/textinputsource/main.c]: Do you count `?` as a valid parameter? I see it treated as `h` at l. 92

File: mac/setup/build.sh (hunk 3 above)
Comment (mcdurdin) [reply mac/setup/build.sh]: I think the build will fail anyway, won't it, with a relevant message? We've got `set -u` at the top of the script.

File: mac/setup/textinputsource/main.c (hunk 4 above)
Comment (mcdurdin) [reply mac/setup/textinputsource/main.c]: Yeah, I didn't write this code so haven't gone through line-by-line for this ☹ I probably should, I guess. For now, I'm okay with this being somewhat incomplete -- the desire is to rewrite the whole installer in a future version, with this being a stopgap.

**Comment by darcywong00** [comment]
Changes seem fine to me, but I'd like to hear how installation goes for @MakaraSok or @jahorton

**Comment by jahorton** [comment]
I haven't done a fresh-VM test, but it seems to have updated my existing installation pretty easily.  Though, the wording still talks about guiding through the install... before it's suddenly done on the next click.

Kinda hard to tell, though - the app just reports build 13.0, with no appended version information to distinguish the actual build.

**Comment by mcdurdin** [comment]
> Kinda hard to tell, though - the app just reports build 13.0, with no appended version information to distinguish the actual build. So I'm kinda assuming that it did perform the actual update.

You should be able to tell the build in the About dialog (available in the language menu once you switch Keyman on).

**Comment by jahorton** [comment]
You mean this one?  Yeah, that's what I was referencing.

Screen Shot 2020-02-12 at 8 57 34 AM

**Comment by mcdurdin** [comment]
Yep, that's it -- there was an issue in that custom build which overwrote the version numbers in the .plist files. I am about to correct that before merging. I see the image below with the released beta version so am comfortable that you are seeing the correct test version.

image

File: mac/Keyman4MacIM/make-km-dmg.sh [hunk 5]
```diff
@@ -139,6 +139,8 @@ elif [[ ! -d "$DEST_DIR" ]]; then
 	fail "Destination dir exists but is not a directory: $2"
 fi
 
+# TODO: Check that no Keyman volume is already mounted.
```
Comment (jahorton) [inline mac/Keyman4MacIM/make-km-dmg.sh]: I assume leaving this a `# TODO` for now is intentional?

File: mac/build.sh [hunk 6]
```diff
@@ -153,7 +155,7 @@ while [[ $# -gt 0 ]] ; do
             ;;
         -deploy-only)
             SKIP_BUILD=true
-            shift # past argument
+            #shift # past argument
```
Comment (jahorton) [inline mac/build.sh]: Is this safe?  Or was this a temp edit that slipped in?

File: mac/Keyman4MacIM/make-km-dmg.sh (hunk 5 above)
Comment (mcdurdin) [reply mac/Keyman4MacIM/make-km-dmg.sh]: yes; I am expecting that we'll stop using .dmg in the future in preference for .pkg, so this won't be necessary to do anyway.

File: mac/build.sh (hunk 6 above)
Comment (mcdurdin) [reply mac/build.sh]: This is safe... it shouldn't have had a `shift` in there so I should probably remove the line altogether.

**Comment by mcdurdin** [comment]
> I wanted to look over the AppleScript components and give that a review, but I'm having trouble finding the source that gives rise to `mac/setup/Install Keyman.app`. I can tell that said file is being used as a _source_ for the actual, final installer:

You have to load the whole app in Apple's Script Editor to view source. It's a bit scary. It is a stopgap, but it may be a little while before we can put together a complete rewrite.

**Comment by mcdurdin** [comment]
image

Looks like I still have some work to do on this. Currently the install app requires 10.15 to start.

**Comment by mcdurdin** [comment]
[keyman-13.0.0.dmg.zip](https://github.com/keymanapp/keyman/files/4202053/keyman-13.0.0.dmg.zip)

This is the latest build with the new way of building the installer. I have successfully tested this on macOS 10.14 Mojave. About to test on 10.7 Lion as well -- when the VM finishes setup.

**Comment by mcdurdin** [comment]
I have rewritten the install app build process and it now builds a compatible installer, and as a bonus the install script is now available in text rather than binary for simpler review.

This installer doesn't work completely on 10.7 but it does on 10.13, 10.14 and 10.15. I am going to leave testing there (out of time at present) -- I think we cease support for 10.7 as it is <1% of users. It should be okay for them to use an older version (if Keyman even worked on those older versions anyway!)

I will update the documentation / home page to note that we have tested on 10.13-10.15 and other versions YMMV.

**Pull request merged**